File: example-crs-webservice/crs-userspace/libs/libMSA/java/libmsa/src/main/java/com/teamatlanta/libmsa/thread/ThreadPool.java
package com.teamatlanta.libmsa.thread;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.ReentrantLock;

public class ThreadPool<T> {

    @FunctionalInterface
    public interface TaskFunction<T> {
        void process(T data);
    }

    private int numThreads;
    private QueuePolicy queuePolicy;
    private TaskFunction<T> func;
    private int numQueues;
    private EventfulQueue<T>[] workQueues;
    private int curQueue = 0;
    private List<Thread> threads = new ArrayList<>();
    private boolean executed = false;

    public enum QueuePolicy {
        GLOBAL, ROUND_ROBIN, BROADCAST
    }

    public ThreadPool(int numThreads, QueuePolicy queuePolicy, TaskFunction<T> func) {
        this.numThreads = numThreads;
        this.queuePolicy = queuePolicy;
        this.func = func;

        this.numQueues = queuePolicy == QueuePolicy.GLOBAL ? 1 : numThreads;
        this.workQueues = new EventfulQueue[numQueues];
        for (int i = 0; i < numQueues; i++) {
            this.workQueues[i] = new EventfulQueue<>();
        }

        createThreads();
    }

    private void createThreads() {
        for (int i = 0; i < numThreads; i++) {
            final int threadId = i; 
            EventfulQueue<T> workQueue = (queuePolicy == QueuePolicy.GLOBAL) ? workQueues[0] : workQueues[i];
            Thread thread = new Thread(() -> worker(threadId, workQueue));
            thread.setDaemon(true);
            threads.add(thread);
        }
    }

    public void enqueue(T data) {
        switch (queuePolicy) {
            case GLOBAL:
                enqueueGlobal(data);
                break;
            case ROUND_ROBIN:
                enqueueRoundRobin(data);
                break;
            case BROADCAST:
                enqueueBroadcast(data);
                break;
            default:
                throw new IllegalStateException("Unexpected value: " + queuePolicy);
        }
    }

    private void enqueueGlobal(T data) {
        workQueues[0].enqueue(data);
    }

    private void enqueueRoundRobin(T data) {
        workQueues[curQueue].enqueue(data);
        curQueue = (curQueue + 1) % numThreads;
    }

    private void enqueueBroadcast(T data) {
        for (int i = 0; i < numThreads; i++) {
            workQueues[i].enqueue(data);
        }
    }

    private void worker(int threadId, EventfulQueue<T> workQueue) {
        ReentrantLock lock = workQueue.getLock();
        Condition condition = workQueue.getCondition();
        while (true) {
            lock.lock();
            try {
                if (workQueue.size() == 0) {
                    condition.await();
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            } finally {
                lock.unlock();
            }

            while (true) {
                T data = workQueue.dequeue();
                if (data == null) {
                    break;
                }
                func.process(data); 
            }
        }
    }

    public void execute() {
        for (Thread thread : threads) {
            thread.start();
        }
        executed = true;
    }

    public void createMoreThreads(int numThreads) {
        if (numThreads <= 0) {
            return;
        }

        if (queuePolicy != QueuePolicy.GLOBAL) {
            EventfulQueue<T>[] temp = new EventfulQueue[this.numQueues + numThreads];
            for (int i = 0; i < this.numQueues; i++){
                temp[i] = workQueues[i];
            }
            workQueues = temp;

            for (int i = this.numThreads; i < this.numThreads + numThreads; i++) {
                workQueues[i] = new EventfulQueue<>();
            }
        }

        for (int i = this.numThreads; i < this.numThreads + numThreads; i++) {
            final int threadId = i; 
            EventfulQueue<T> workQueue = (queuePolicy == QueuePolicy.GLOBAL) ? workQueues[0] : workQueues[i];
            Thread thread = new Thread(() -> worker(threadId, workQueue));
            thread.setDaemon(true);
            threads.add(thread);
            if (executed) {
                thread.start();
            }
        }

        this.numThreads += numThreads;
        if (queuePolicy != QueuePolicy.GLOBAL) {
            this.numQueues += numThreads;
        }
    }

    public int getNumWorkQueues(){
        return workQueues.length;
    }

    public EventfulQueue<T>[] getWorkQueues(){
        return workQueues;
    }
}

File: example-crs-webservice/crs-userspace/libs/libMSA/go/thread/threadpool.go
package thread

import (
	//"fmt"
)

type QueuePolicy int

const (
	GLOBAL QueuePolicy = iota
	ROUND_ROBIN
	BROADCAST
)

type ThreadPool struct {
	numThreads   int
	queuePolicy  QueuePolicy
	taskFunction func(data any)
	numQueues    int
	workQueues   []*EventfulQueue[any]
	curQueue     int
	threads      []*threadWorker
	executed     bool
}

type threadWorker struct {
	id        int
	workQueue *EventfulQueue[any]
}

func NewThreadPool(numThreads int, queuePolicy QueuePolicy, taskFunction func(data any)) *ThreadPool {
	tp := &ThreadPool{
		numThreads:   numThreads,
		queuePolicy:  queuePolicy,
		taskFunction: taskFunction,
		curQueue:     0,
		executed:     false,
	}

	if queuePolicy == GLOBAL {
		tp.numQueues = 1
		tp.workQueues = []*EventfulQueue[any]{NewEventfulQueue[any]()}
	} else {
		tp.numQueues = numThreads
		tp.workQueues = make([]*EventfulQueue[any], numThreads)
		for i := 0; i < numThreads; i++ {
			tp.workQueues[i] = NewEventfulQueue[any]()
		}
	}

	tp.createThreads()

	return tp
}

func (tp *ThreadPool) createThreads() {
	for i := 0; i < tp.numThreads; i++ {
		workQueue := tp.workQueues[0]
		if tp.queuePolicy != GLOBAL {
			workQueue = tp.workQueues[i]
		}

		worker := &threadWorker{id: i, workQueue: workQueue}
		tp.threads = append(tp.threads, worker)
	}
}


func (tp *ThreadPool) Enqueue(data any) {
	switch tp.queuePolicy {
		case GLOBAL:
			tp.workQueues[0].Enqueue(data)
		case ROUND_ROBIN:
			tp.workQueues[tp.curQueue].Enqueue(data)
			tp.curQueue = (tp.curQueue + 1) % tp.numQueues
		case BROADCAST:
			for _, queue := range tp.workQueues {
				queue.Enqueue(data)
			}
	}
}

func (tp *ThreadPool) Start() {
	for _, worker := range tp.threads {
		go worker.run(tp.taskFunction)
	}
	tp.executed = true
}

func (tw *threadWorker) run(taskFunction func(data any)) {
	lock := tw.workQueue.GetLock()
	cond := tw.workQueue.GetCond()
	for {
		lock.Lock()
		if tw.workQueue.queue.Len() == 0{
			cond.Wait()
		}
		lock.Unlock()

		for {
			event := tw.workQueue.Dequeue()
			if event == nil {
				break
			}
			taskFunction(*event)
		}
	}
}

func (tp *ThreadPool) AddThreads(numThreads int) {
	if numThreads <= 0 {
		return
	}

	for i := tp.numThreads; i < tp.numThreads+numThreads; i++ {
		var workQueue *EventfulQueue[any]
		if tp.queuePolicy != GLOBAL {
			workQueue = NewEventfulQueue[any]()
			tp.workQueues = append(tp.workQueues, workQueue)
		} else {
			workQueue = tp.workQueues[0]
		}

		worker := &threadWorker{id: i, workQueue: workQueue}
		tp.threads = append(tp.threads, worker)

		if tp.executed {
			go worker.run(tp.taskFunction)
		}
	}

	tp.numThreads += numThreads
	if tp.queuePolicy != GLOBAL {
		tp.numQueues += numThreads
	}
}